File: Shapes.py
#Josef Los
#1/31/19
#Homework 1
#A simple program that deals with OOP concepts

class Shape:
	def __init__(self, name, sides=-1):
		self.name = name
		self.sides = sides
	
	def display(self):
		print "Hello! My name is " + self.name + ". I am a " + self.__class__.__name__ + " and I have " + str(self.sides) + " sides!"
	
	def getname(self):
		return self.name

	def setname(self, newname):
		self.name = newname

	def getsides(self):
		return self.sides

	def __cmp__(self, other):
		return self.name != other


class Circle(Shape):
	def __init__(self, name):
		self.name = name
		self.sides = 0

class Rectangle(Shape):
	def __init__(self, name):
		self.name = name
		self.sides = 4

class Triangle(Shape):
	def __init__(self, name):
		self.name = name
		self.sides = 3

class ShapeDatabase:
	def __init__(self, shapes):
		self.shapes = shapes
	
	def add(self, shape):
		self.shapes.append(shape)
	
	def remove(self, shape):
		self.shapes.remove(shape)

	def listshapes(self):
		for i in self.shapes:
			i.display()

	def sortbysides(self):
		self.shapes.sort(key=lambda shape: shape.sides)

	def sortbyname(self):
		self.shapes.sort(key=lambda shape: shape.name)

	def getsize(self):
		return len(self.shapes)
	
			
#initialize the shapes and the database
mycircle = Circle("Terry")	
mytriangle = Triangle("Lisa")
myrectangle = Rectangle("Bob")
myshapes = [mycircle, mytriangle, myrectangle]
mydatabase = ShapeDatabase(myshapes)

#print the size of database and have each shape display itself
print mydatabase.getsize()
mydatabase.listshapes()

File: shapes.cpp
//Function definitions for the various shapes

#include <iostream>
#include <string>
#include "shapes.h"
#include "display.h"

std::string Shape::getName()
{
	return name;
}

int Shape::getSides()
{
	return sides;
}

bool Shape::operator < (const Shape& other) const
{
	return (sides < other.sides);
}

Circle::Circle(std::string n)
//: name(n), sides(0)//, displayMethod(Display())
{
	name = n;
	sides = 0;
	displayMethod = Display();
}

Triangle::Triangle(std::string n)
//: name(n), sides(3)//, displayMethod(Display())
{
	name = n;
	sides = 3;
	displayMethod = Display();
}

Square::Square(std::string n)
//: name(n), sides(4), displayMethod(Display())
{
	name = n;
	sides = 4;
	displayMethod = Display();
}

void Circle::display()
{
	displayMethod.display(name, sides);	
}

void Triangle::display()
{
	displayMethod.display(name, sides);	
}

void Square::display()
{
	displayMethod.display(name, sides);	
}
